### PR TITLE
feat(auth): add auth provider for Bitbucket Server

### PR DESCRIPTION
**Pull request by KatharinaSick**
Signed-off-by: Katharina Sick <katharina.sick@dynatrace.com>

## Hey, I just made a Pull Request!

I added an Auth Provider to authenticate against Bitbucket Server. Unfortunately, no PassportJS strategy for Bitbucket Server exists, therefore I used the [passport-oauth2](http://www.passportjs.org/packages/passport-oauth2/) strategy.

Screenshots:
providers
signin
consent

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by KatharinaSick** [comment]
I just checked why the CI/verify workflows are failing: `yarn prettier:check` shows a warning for file `.changeset/slow-moles-sin.md`. This hasn't been created/touched by me -> should I still fix it?

File: plugins/auth-backend/api-report.md [hunk 1]
```diff
@@ -480,6 +493,23 @@ export const providers: Readonly<{
...
+            signIn?:
+              | {
+                  resolver: SignInResolver<BitbucketServerOAuthResult>;
+                }
+              | undefined;
+          }
+        | undefined,
+    ) => AuthProviderFactory;
+    resolvers: Readonly<{
+      emailMatchingUserEntityProfileEmail: () => SignInResolver<unknown>;
```
Comment (freben) [inline plugins/auth-backend/api-report.md:510]: I thought this one was going to say `SignInResolver<BitbucketServerOAuthResult>`.

File: plugins/auth-backend/src/providers/bitbucketServer/provider.test.ts [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+          ],
+        },
+        avatarUrl: avatarUrl,
+      }),
+    });
+  }
+};
+
+describe('BitbucketServerAuthProvider', () => {
+  const originalFetch = global.fetch;
```
Comment (freben) [inline plugins/auth-backend/src/providers/bitbucketServer/provider.test.ts:98]: We pervasively make use of `msw` to stub out network traffic, making things fast and robust. It also protects against accidentally leaking real traffic outside of the test.

https://github.com/backstage/backstage/blob/3cf2190c2df803833b7f6f057d873bb6e96bcb46/plugins/scaffolder/src/api.test.ts#L28

**Review by freben — COMMENTED** [review]
Alright I think this looks good to me! @Rugvip any additional comments?

File: plugins/auth-backend/api-report.md (hunk 1 above)
Comment (KatharinaSick) [reply plugins/auth-backend/api-report.md:510]: Should/can I do something about this?

Comment (freben) [reply plugins/auth-backend/api-report.md:510]: you can declare the type explicitly like github does here https://github.com/backstage/backstage/blob/01a075ec1df4d493b94df660ecc5660fd0837898/plugins/auth-backend/src/providers/github/provider.ts#L331

File: plugins/auth-backend/src/providers/bitbucketServer/provider.test.ts (hunk 2 above)
Comment (KatharinaSick) [reply plugins/auth-backend/src/providers/bitbucketServer/provider.test.ts]: Done. I never worked with `msw` before, so I hope everything is fine. 

Btw, thanks for the suggestion - `msw` looks like a great tool :)

File: plugins/auth-backend/package.json [hunk 3]
```diff
@@ -78,6 +78,7 @@
   "devDependencies": {
     "@backstage/backend-test-utils": "workspace:^",
     "@backstage/cli": "workspace:^",
+    "@testing-library/jest-dom": "^5.16.5",
```
Comment (Rugvip) [inline plugins/auth-backend/package.json:81]: This should not be added to a backend package

File: plugins/auth-backend/package.json [hunk 4]
```diff
@@ -89,6 +90,7 @@
     "@types/passport-saml": "^1.1.3",
     "@types/passport-strategy": "^0.2.35",
     "@types/xml2js": "^0.4.7",
+    "cross-fetch": "^3.1.5",
```
Comment (Rugvip) [inline plugins/auth-backend/package.json:93]: Use `node-fetch` instead, since this is a backend package

**Review by Rugvip — CHANGES_REQUESTED** [review]
Thank you! 👍 

Just a small nit

File: plugins/auth-backend/package.json (hunk 4 above)
Comment (KatharinaSick) [reply plugins/auth-backend/package.json:93]: I tried to remove it, but somehow `msw` fails to mock my requests without [this line](https://github.com/backstage/backstage/pull/16186/files#diff-31f8175f56ceca064e3dd7177f22c8d1366d85cb01021d5d80d56864e5771816R30)

Comment (KatharinaSick) [reply plugins/auth-backend/package.json:93]: nvm - I just figured it out :)

File: plugins/auth-backend/src/providers/bitbucketServer/provider.ts [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+    };
+  }
+
+  private async fetchProfile(
+    result: BitbucketServerOAuthResult,
+  ): Promise<PassportProfile> {
+    // Get current user name
+    let whoAmIResponse;
+    try {
+      whoAmIResponse = await fetch(
```
Comment (Rugvip) [inline plugins/auth-backend/src/providers/bitbucketServer/provider.ts:182]: You'll want to `import fetch from 'node-fetch'`, it's not available everywhere yet

File: plugins/auth-backend/src/providers/bitbucketServer/provider.test.ts [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+import {
+  bitbucketServer,
+  BitbucketServerAuthProvider,
+  BitbucketServerOAuthResult,
+} from './provider';
+import { setupServer } from 'msw/node';
+import { setupRequestMockHandlers } from '@backstage/test-utils';
+import { rest } from 'msw';
+
+global.fetch = require('node-fetch');
```
Comment (Rugvip) [inline plugins/auth-backend/src/providers/bitbucketServer/provider.test.ts:29]: Related to the other comment, this shouldn't be needed.

**Review by Rugvip — CHANGES_REQUESTED** [review]
Just a small tweak around the usage of `node-fetch`. Fetch isn't stable API in Node just yet 😉

File: plugins/auth-backend/src/providers/bitbucketServer/provider.ts (hunk 5 above)
Comment (KatharinaSick) [reply plugins/auth-backend/src/providers/bitbucketServer/provider.ts:183]: Fixed - thanks for your help! :) 
As a person who is doing mostly devops and backend stuff, those things are quite hard for me to get started with

**Pull request merged**